pkg/googlegroups/groupreader.go

```diff
@@ -0,0 +1,116 @@
+)
+
+const (
+	MemberTypeUser  = "USER"
+	MemberTypeGroup = "GROUP"
+)
+
+// Ensure we conform to the interface.
+var _ groupsync.GroupReader = (*GroupReader)(nil)
+
+// GroupReader provides read operations for groups and users in GCP.
+type GroupReader struct {
+	identity *cloudidentity.Service
+	admin    *admin.Service
+}
+
+// NewGroupReader create a new GroupReader.
+func NewGroupReader(identityService *cloudidentity.Service, adminService *admin.Service) *GroupReader {
+	return &GroupReader{
+		identity: identityService,
+		admin:    adminService,
+	}
+}
+
+// Descendants retrieve all users (children, recursively) of a group.
+func (g GroupReader) Descendants(ctx context.Context, groupID string) ([]*groupsync.User, error) {
+	var members []*groupsync.User
+	err := g.identity.Groups.Memberships.SearchTransitiveMemberships(groupID).Context(ctx).Pages(ctx,
+		func(page *cloudidentity.SearchTransitiveMembershipsResponse) error {
+			for _, m := range page.Memberships {
+				// we only want user memberships but the API doesn't give us any type information.
+				// Instead, we infer the type from the resource name. e.g. groups have a resource
+				// name of the form `groups/%s` and users of the form 'users/%d'
+				if strings.HasPrefix(m.Member, "users/") {
+					members = append(members, &groupsync.User{ID: m.Member})
+				}
+			}
+			return nil
+		},
+	)
+	return members, err //nolint:wrapcheck
+}
+
+// GetGroup retrieves the Group with the given ID. The ID must be of the form `groups/{group}`.
+func (g GroupReader) GetGroup(ctx context.Context, groupID string) (*groupsync.Group, error) {
+	group, err := g.identity.Groups.Get(groupID).Context(ctx).Do()
+	if err != nil {
+		return nil, fmt.Errorf("could not get group: %w", err)
+	}
+	return &groupsync.Group{
+		ID:         group.Name,
+		Attributes: group,
+	}, nil
+}
+
+// GetMembers retrieves the direct members (children) of the group with given ID.
+// This includes both users and subgroups.
+func (g GroupReader) GetMembers(ctx context.Context, groupID string) ([]groupsync.Member, error) {
+	var members []groupsync.Member
+	logger := logging.FromContext(ctx)
+	if err := g.identity.Groups.Memberships.List(groupID).Context(ctx).Pages(ctx,
+		func(page *cloudidentity.ListMembershipsResponse) error {
+			for _, m := range page.Memberships {
+				if m.Type == MemberTypeGroup {
+					members = append(members, &groupsync.GroupMember{Grp: &groupsync.Group{ID: m.PreferredMemberKey.Id}})
+				} else if m.Type == MemberTypeUser {
+					members = append(members, &groupsync.UserMember{Usr: &groupsync.User{ID: m.PreferredMemberKey.Id}})
+				} else {
+					logger.WarnContext(ctx, "unrecognized member type encountered",
+						"group_id", groupID,
+						"member", m,
+					)
+				}
+			}
+			return nil
+		},
+	); err != nil {
+		return nil, fmt.Errorf("could not get group members: %w", err)
+	}
+	return members, nil
+}
+
+// GetUser retrieves the User with the given ID. Should be of the form.
```

Comment:
This comment looks unintentionally truncated.

Reply:
fixed.